cmd/sovereignnode/dataCodec/dataCodec.go

```diff
@@ -0,0 +1,482 @@
+		for i, arg := range eventData.Args {
+			arguments[i] = abi.BytesValue{Value: arg}
+		}
+		args = abi.InputListValue{
+			Items: arguments,
+		}
+	} else {
+		gasLimit = nil
+		function = nil
+		args = nil
+	}
+
+	eventDataStruct := abi.StructValue{
+		Fields: []abi.Field{
+			{
+				Name:  "tx_nonce",
+				Value: abi.U64Value{Value: eventData.Nonce},
+			},
+			{
+				Name: "gas_limit",
+				Value: abi.OptionValue{
+					Value: gasLimit,
+				},
+			},
+			{
+				Name: "function",
+				Value: abi.OptionValue{
+					Value: function,
+				},
+			},
+			{
+				Name: "args",
+				Value: abi.OptionValue{
+					Value: args,
+				},
+			},
+		},
+	}
+
+	encodedOp, err := dc.Serializer.Serialize([]any{eventDataStruct})
+	if err != nil {
+		return nil, err
+	}
+
+	operationBytes, _ := hex.DecodeString(encodedOp)
+	return operationBytes, nil
+}
+
+// DeserializeEventData will deserialize bytes to an event data structure
+func (dc *DataCodec) DeserializeEventData(data []byte) (*sovereign.EventData, error) {
+	if len(data) == 0 {
+		return nil, fmt.Errorf("empty data provided")
+	}
+
+	eventDataStruct := abi.StructValue{
+		Fields: []abi.Field{
+			{
+				Name:  "tx_nonce",
+				Value: &abi.U64Value{},
+			},
+			{
+				Name: "gas_limit",
+				Value: &abi.OptionValue{
+					Value: &abi.U64Value{},
+				},
+			},
+			{
+				Name: "function",
+				Value: &abi.OptionValue{
+					Value: &abi.BytesValue{},
+				},
+			},
+			{
+				Name: "args",
+				Value: &abi.OptionValue{
+					Value: &abi.OutputListValue{
+						ItemCreator: func() any { return &abi.BytesValue{} },
+					},
+				},
+			},
+		},
+	}
+
+	err := dc.Serializer.Deserialize(hex.EncodeToString(data), []any{&eventDataStruct})
+	if err != nil {
+		return nil, err
+	}
+
+	nonce := eventDataStruct.Fields[0].Value.(*abi.U64Value).Value
+
+	gasLimit := uint64(0)
+	optionGas := eventDataStruct.Fields[1].Value.(*abi.OptionValue).Value
+	if optionGas != nil {
+		gasLimit = optionGas.(*abi.U64Value).Value
+	}
+
+	function := make([]byte, 0)
+	optionFunc := eventDataStruct.Fields[2].Value.(*abi.OptionValue).Value
+	if optionFunc != nil {
+		function = optionFunc.(*abi.BytesValue).Value
+	}
+
+	args := make([][]byte, 0)
+	optionArgs := eventDataStruct.Fields[3].Value.(*abi.OptionValue).Value
+	if optionArgs != nil {
+		items := optionArgs.(*abi.OutputListValue).Items
+		if items != nil && len(items) > 0 {
+			for _, item := range items {
+				arg := item.(*abi.BytesValue).Value
+				args = append(args, arg)
+			}
+		}
+	}
+
+	var td *sovereign.TransferData
+	if len(function) == 0 {
+		td = nil
+	} else {
+		td = &sovereign.TransferData{
+			GasLimit: gasLimit,
+			Function: function,
+			Args:     args,
+		}
+	}
+
+	return &sovereign.EventData{
+		Nonce:        nonce,
+		TransferData: td,
+	}, nil
+}
+
+// SerializeTokenData will receive an esdt token data and serialize it
+func (dc *DataCodec) SerializeTokenData(tokenData sovereign.EsdtTokenData) ([]byte, error) {
```

Comment:
I would say for now we can use it like this because is custom data serialized/deserialized and is only for sovereign bridge operations.
Also, (for example) to complete one outgoing operation you would need to deserialize TokenData, deserialize EventData and SerializeOperation...

Anyways, in the future this codec component will be moved out of the node. The nod will receive/send only bytes and will not be responsible with the serialization.

Reply:
ok